libs/service/service.go

```diff
@@ -52,6 +40,18 @@ type Service interface {
 	Wait()
 }
 
+// ServiceImplementation describes the implementation that the
+// BaseService implementation wraps.
+type ServiceImplementation interface { // nolint:revive
+	Service
+
+	// Called by the Services Start Method
```

Comment:
I think see what's intended here, but I am not yet convinced this is an improvement over the single, larger interface. Is this intended to be a step toward a later change where these "hook" methods go away?

An implementation still has to implement more or less the same methods as before, but there's no single place where we document how the implementation is expected to behave. That isn't this PR's fault: It wasn't documented before either, but all the methods were in one place. At minimum, I feel like we could at least improve the comments.

But more generally, can we get rid of OnStart entirely? I think that existed mainly because the silly BaseService wrapper hid the Start plumbing from the implementation. Can we get rid of that? Is that the intent toward which this is a step? (If so I'm fine not investing too heavily in polish yet)

Reply:
I think this is a step in this direction: this one was more about getting the contexts through and less about changing the hook methods, which I'd like to do that later.

Reply:
Deferring deeper cuts to later makes sense. But what does the target state look like? Can we hope to reach a state where the "implementation" interface goes away entirely, and the concrete implementations just implement `Service` directly?

If that's the goal, I'm enthusiastic, and won't worry too much about the intermediate steps (apart perhaps from maybe requesting some TODO comments and an issue 🙂).

Reply:
I think the target time frame for this is somewhat longer than is ideal (because we need/want to get rid of a lot of things that _are_ service implementations but shouldn't be (connections.) 

Having thought about the end stage for all of 5 minutes, I think I kind of want `service.Service` to become concrete, and the interface to be the thing that's currently `Implementation` (except with like... reasonable names).

Reply:
Hm. I generally like concrete types, but I'm not sure I see the endgame for the body of the implementation in that case: A concrete `Service` could gain a `Run` function pointer (in that sense, similar to the way CLI packages like Cobra work), but that meshes weirdly with the interface unless we also plumb the implementation receiver in (e.g., via the context or another argument).

The idea of splitting the interface into "plumbing" and "implementation" is a good one, but I think we'd have to decide what the implementation protocol is. Notionally, the "init-run-cleanup" pattern, which is sort of what we have here, is reasonable. Following that idea, `Service` could become concrete and `Implementation` could become something like:

```go
type Implementation interface{
  Start(ctx context.Context) error  // non-blocking; failure means the server can't start
  Stop(ctx context.Context)         // non-blocking; signal to stop only
  Wait(ctx context.Context) error   // wait for shutdown and return error
}
```

This would give a concrete Server enough to operate on.

I think we could get to that from where we are now, but I think we will have to figure out how to disentangle the weird Enterprise Java style nested structure we're using now.

Not stuff we have to do here, just probing the design space a little.

Reply:
Yeah, I realize that the thing I dislike the most about the state (without this) is that there's a sort of almost-cylce (e.g that all services all must use base service, and base service must have a pointer to the implementation. 

This doesn't actually fix that, but it does sort of say "the thing that base service needs isn't base service itself. 

Why not make Start/Stop blocking, drop the wait (or leave it for the executor/outer implementation). 

While I think we should pass a context to wait, it's hard because you need a context with a different root during shutdown (if you use the SIGTERM to cancel the context,) but making it a context that is never canceled is also hard. 

----

It's probably the case, however, that the _logger_ and not start/stop/wait orchestration is going to be slightly harder to pull apart, but that's a different PR.

Reply:
> Why not make Start/Stop blocking

Start could block—but then there's no easy way to report "initialization" errors: If the caller runs Start in a goroutine, and the server fails to start, nobody will know until it shuts down and reports failure. This is a case where I think it makes sense for the server to maintain its own concurrency.

I don't think Stop should ever block, because the code that wants to tell the service to stop (e.g., the node getting ready to shut down) is not necessarily in the same place as the code that needs to wait for it.  Moreover, more than one piece of code may need to wait for the service (e.g., A depends on B, so after stopping A, both you and A want to wait for B). IMO "Stop" should be a straightforward signal to the server.

> drop the wait (or leave it for the executor/outer implementation).

That's an option, but I think it isn't the right choice, for the reasons I mentioned above.

> While I think we should pass a context to wait, it's hard because you need a context with a different root during shutdown

Don't read too much into my demo interface. My intent was that the context passed to wait is not governing the service lifecycle, but the Wait call: If the caller wants to wait a while but then give up, they would pass in a timeout context to Wait. But that is kind of optional, and I don't think it's essential.

Reply:
I think it's fine if start blocks, but only does initialization and doesn't block until the service exits, and we can kind of document that. 

I think each service's implementation of Stop can block, and the _executor_ can *not* block, which means that every service doesn't need to implement it's own Wait functionality, which seems easy for implementors to get wrong, and I guess my goal is make the thing that implementors have to actually implement very small and very difficult to get wrong.

Reply:
> think it's fine if start blocks, but only does initialization and doesn't block until the service exits

Yeah, that's what I meant. Not that it can't do work, only that it doesn't wait while the service runs.

> I think each service's implementation of Stop can block

Sure, that's a perfectly reasonable implementation. But then there's still this question of how do I wait for a service I don't want to stop? I think we already have cases of that, where one service depends on another.

We could have Wait _also_ block for the service to end without triggering stop, but then we have the same logic in two methods. Conversely, if we want to tell something to shut down without immediately waiting, we'd have to put that in a goroutine, and then wait again later. So the Stop/Wait distinction isn't really adding new mass to the interface, it's just pulling apart two concerns that already exist.

That said: What you're describing is a perfectly viable interface.

Reply:
> > think it's fine if start blocks, but only does initialization and doesn't block until the service exits
> 
> Yeah, that's what I meant. Not that it can't do work, only that it doesn't wait while the service runs.
> 
> > I think each service's implementation of Stop can block
> 
> Sure, that's a perfectly reasonable implementation. But then there's still this question of how do I wait for a service I don't want to cause to stop by virtue of waiting? I think we already have cases of that, where one service depends on another.
> 
> We could have Wait _also_ block for the service to end without triggering stop, but then we have the same logic in two methods (it could be pulled out and shared, but the two methods would still share a concern). Conversely, if we want to tell something to shut down without immediately waiting, we'd have to put that in a goroutine, and then wait again later. So the Stop/Wait distinction isn't really adding new mass to the interface, it's just pulling apart two concerns that already exist.
> 
> That said: What you're describing is a perfectly viable interface.

Reply:
I don't see that we'd have a situation where the service.Service wrapper would have the semantics that you expect, but that the implementations would have the developer ergonomics that I'm lobbying for, and I think it all works pretty well? 

Anyway, we can get there in another PR or 12.

Reply:
Reading over these comments, I do feel like we should work on having a defined endgame and I think the design thinking around how the node service should look like (and the other services for that matter) deserves it's own ADR even if it's small to start with and builds up once we develop a better understanding. I just feel that such substantial changes need some artifact others can later refer to.

Reply:
I think the current change is a good one, and the difference between our positions here feels *very* narrow and equally possible given this change.

Reply:
I have no objections to this change. My suggestions above were not meant to be blocking.

Reply:
```suggestion
	// Called by the BaseService Start implementation.
```

Reply:
Yeah my comment was aimed more at the broader picture not this specific PR